pkcs11/pkcs11.go

```diff
@@ -1687,3 +1687,117 @@ func (s *Slot) generateECDSA(o keyOptions) (crypto.PrivateKey, error) {
 	}
 	return priv, nil
 }
+
+func (r *rsaPrivateKey) Decrypt(_ io.Reader, encryptedData []byte, opts crypto.DecrypterOpts) ([]byte, error) {
+	if o, ok := opts.(*rsa.OAEPOptions); ok {
+		return r.decryptOAEP(encryptedData, o)
+	}
+
+	cEncDataBytes := toCBytes(encryptedData)
+
+	m := C.CK_MECHANISM{C.CKM_RSA_PKCS, nil, 0}
+
+	rv := C.ck_decrypt_init(r.o.fl, r.o.h, &m, r.o.o)
+	if err := isOk("C_DecryptInit", rv); err != nil {
+		return nil, err
+	}
+
+	var cDecryptedLen C.CK_ULONG
+
+	// First call is used to determine length necessary to hold decrypted data (PKCS #11 5.2)
+	rv = C.ck_decrypt(r.o.fl, r.o.h, &cEncDataBytes[0], C.CK_ULONG(len(cEncDataBytes)), nil, &cDecryptedLen)
+	if err := isOk("C_Decrypt", rv); err != nil {
+		return nil, err
+	}
+
+	cDecrypted := make([]C.CK_BYTE, cDecryptedLen)
+
+	rv = C.ck_decrypt(r.o.fl, r.o.h, &cEncDataBytes[0], C.CK_ULONG(len(cEncDataBytes)), &cDecrypted[0], &cDecryptedLen)
+	if err := isOk("C_Decrypt", rv); err != nil {
+		return nil, err
+	}
+
+	decrypted := toBytes(cDecrypted)
+
+	// Removes null padding (PKCS#11 5.2): http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/os/pkcs11-base-v2.40-os.html#_Toc416959738
+	decrypted = bytes.Trim(decrypted, "\x00")
+
+	return decrypted, nil
+}
+
+func (r *rsaPrivateKey) decryptOAEP(encryptedData []byte, opts *rsa.OAEPOptions) ([]byte, error) {
```

Comment:
nit: There is a lot of shared logic between this and the `Decrypt` function.

My suggestion:

Spit algorithm specific parts into helper functions, and then keep the shared logic in `Decrypt`.

It seems creating the `cParam` object is the only distinct part of each algorithm. 

What do you think?

Reply:
I agree, creating the mechanism to pass into C_DecryptInit and C_Decrypt is the only difference between the two functions. Would it be okay to do this logic in an if statement instead of a helper function as the function will only be called once?

Reply:
My 2c is to have the if statement and the mechanism constructor in that helper function. That should strike a nice balance for now. More complicated mechanisms may get verbose

Reply:
I think I'm running into a similar situation: https://github.com/google/go-pkcs11/pull/15#discussion_r1309074334.

Creating a cParam in a helper function makes whoever calls the function responsible for freeing the malloc. Should I follow the same pattern and free it after making a call to the helper function?

Reply:
Ah gotcha. Yea go ahead and just keep all the logic in the main function. It can be cleaned up later

Reply:
Okay, I just updated the main Decrypt function to handle both cases and removed the decryptOAEP function.